core/src/main/java/org/mule/runtime/core/policy/CompositeSourcePolicy.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.policy;
+
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.core.api.DefaultMuleException;
+import org.mule.runtime.core.api.Event;
+import org.mule.runtime.core.api.policy.SourcePolicyParametersTransformer;
+import org.mule.runtime.core.api.processor.Processor;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * {@link SourcePolicy} created from a list of {@link ParameterizedPolicy}.
+ * <p>
+ * Takes care of chaining the list of {@link ParameterizedPolicy} to create a single
+ * policy that can be applied to a source.
+ *
+ * @since 4.0
+ */
+public class CompositeSourcePolicy implements SourcePolicy {
+
+  private List<ParameterizedPolicy> parameterizedPolicies;
+  private Optional<SourcePolicyParametersTransformer> sourcePolicyParametersTransformer;
+  private PolicyStateHandler policyStateHandler;
+
+  /**
+   * Creates a new composite policy.
+   *
+   * @param parameterizedPolicies list of {@link ParameterizedPolicy} to chain together.
+   * @param sourcePolicyParametersTransformer transformer from the response parameters to a message and vice versa.
+   * @param policyStateHandler state handler for policies execution.
+   */
+  public CompositeSourcePolicy(List<ParameterizedPolicy> parameterizedPolicies,
+                               Optional<SourcePolicyParametersTransformer> sourcePolicyParametersTransformer,
+                               PolicyStateHandler policyStateHandler) {
+    this.parameterizedPolicies = parameterizedPolicies;
+    this.sourcePolicyParametersTransformer = sourcePolicyParametersTransformer;
+    this.policyStateHandler = policyStateHandler;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Event process(Event sourceEvent, Processor nextOperation,
+                       MessageSourceResponseParametersProcessor messageSourceResponseParametersProcessor)
+      throws Exception {
+    return new NextOperationCall(sourceEvent, nextOperation, messageSourceResponseParametersProcessor).process(sourceEvent);
+  }
+
+  /**
+   * Inner class that implements the actually chaining of policies.
+   */
+  public class NextOperationCall implements Processor {
+
+    private int index = 0;
+    private Processor nonPolicyNextOperation;
+    private Event sourceEvent;
+    private MessageSourceResponseParametersProcessor messageSourceResponseParametersProcessor;
```

Comment:
Right, but don you think that there is enough duplicate code to try to generalize it?